### DISCUSSION TITLE
An alternative to requiring pull requests to be "up to date" with their target

### DISCUSSION THREAD

**Discussion by eyalroth** [community/community#7399] [Pull Requests]
If I understand correctly, the branch protection rule feature [Require branches to be up to date before merging](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-status-checks-before-merging) is aimed at making sure that a PR is tested with the latest code, so that when it is merged it will not break the target branch.

While this feature helps at making sure that PRs don't break their targets, it does come with its own problems, and I would like to suggest an alternative that does not exhibit these problems while still keeping the original goal in mind.

**What is the issue with "up to date" checks?**

The feature works sufficiently well when the new changes to the target branch conflict with the PR changes, or when those two do not conflict technically but will break the codebase if merged; however, it's completely redundant in the cases when the PR changes do not interact in any manner with the new changes to the target branch.

I would imagine that these situations are more prevalent the larger the codebase is, but even as I currently work in a micro-services oriented organization, with small repositories and with very few developers working on the same repository at once, I frequently encounter these situations in which my PR needs to be updated despite it being fully compatible with the newest changes to the target branch.

In these situations, where the PR changes and the new changes on the target branch are compatible, two major problems arise with the feature:

1. It requires an additional manual step before merging -- either updating the PR via the UI with a merge-commit or rebasing the PR commits on top of the new target branch head (with a force push). This is the lesser of the two problems, and might be mitigated with the introduction of [PR merge queue](https://github.blog/changelog/2021-10-27-pull-request-merge-queue-limited-beta/).

2. The bigger problem is not with the requirement for manual intervention, but with the methods that this features expects the PR to be "resolved" -- merge-commit or rebase with force push. The former method introduces many meaningless merge-commits that (a) pollute the overall history of the repository and (b) makes interactive rebases on the PR branch harder, while the latter method requires a force-push that which affects any other branch targeting the PR's source branch.

These problems become a real pain whenever I split my work into multiple smaller PRs that build on top of each other (a "PR train").

**What is the alternative?**

Create a check on the PR that instead of running the tests on the PR's branch, it will:
1. Switch to the _target_ branch of the PR.
2. Locally merge-commit the source branch of the PR to the target branch; this could obviously fail if there are conflicts.
3. Run the tests.

Such a check satisfies the original goal of making sure that a PR doesn't break the target branch, while requiring no human intervention nor adds redundant merge-commits in the many cases where the PR is completely compatible with the new changes to the target branch.

The only thing that is required to make this alternative feature possible is adding a webhook on PRs that is triggered whenever a PR's target branch is updated. I've [recently asked](https://github.community/t/trigger-workflow-on-a-pull-request-when-the-base-target-branch-is-updated/209470) about this capability, and seen that others have been asking for it [here](https://github.community/t/trigger-a-workflow-when-the-base-reference-of-a-pull-request-is-updated/117299/3) and [here](https://stackoverflow.com/questions/67868483/how-to-rerun-github-actions-workflow-when-the-pull-request-base-branch-updated). I'd imagine that the webhook exists -- otherwise how does the "up to date" feature works in the first place?

**Comment by torressam333** [comment]
Did you ever find a solution. My team is having major headaches with your first point about non conflicting changes requiring "Updating" their pr with a rebase.
